### ISSUE TITLE
[Fatal Error] Random Access Violation Crashes

### ISSUE THREAD

**Issue by dfdffish** [xStarbound/xStarbound#9]
During a recent playthrough with my friend, I encountered a severe issue that led to the crash of my xserver with an **"Access Violation"** error message while progressing through the **"apexmission1"**.

Initially, I suspected the problem might be related to my modpack, so I attempted to restart the server to resume our game. However, I soon discovered that **the crash occurs randomly** throughout the game.

The crashes are indeed unpredictable. I've experienced them at various points within the same "InstanceWorld".

Upon reviewing the logs, the information provided is not clear to me, making it difficult to identify the root cause.

This led me to consider whether the issue might be a bug with the xserver. To test this hypothesis, I conducted experiments in a controlled environment without any mods.

I executed the command "/warp InstanceWorld:apexmission1" and attempted to progress through the level multiple times to see if I could consistently reproduce the error. The results were **inconsistent**; sometimes the game functioned without issue, while at other times it crashed, particularly at the beginning of the level near NPC Lana Blake or during gameplay progression.

Unfortunately, I have not been able to find a reliable method to trigger this "Access Violation" error, although it seems to occur **more frequently in the "apexmission1"**, but not every time.

Additionally, I've noticed that when the crash occurs, it only affects the xserver, causing the xclients to disconnect and return to the main menu. But I do recall experiencing similar crashes in xclient singleplayer, which resulted in the entire game crashing. However, I do not have logs to prove these.

I hope I'm not mistaken anything.   :)

[xserver(Clean).log](https://github.com/user-attachments/files/16139818/xserver.Clean.log)
[xserver.log](https://github.com/user-attachments/files/16139820/xserver.log)

**Comment by FezzedOne** [comment]
Could you download the PDB files and put `xclient.pdb` and `xserver.pdb` into your `win64\` folder? The logs are *useless* without those. Also, to make debugging easier, you can use `/warp InstanceWorld:apexmission1:-:-` to quickly warp to the apex mission. You must have `/admin` enabled to use that command.

**Comment by dfdffish** [comment]
Okay sure. I'll try if I can reproduce again with these pdb files.

**Comment by dfdffish** [comment]
It did show something more.
[xclient.log](https://github.com/user-attachments/files/16140553/xclient.log)
[xserver.log](https://github.com/user-attachments/files/16140554/xserver.log)

**Comment by FezzedOne** [comment]
You or your friend didn't put `xclient.pdb` in the correct place on an xClient install. However, I can at least use the xServer log now. Make sure your friend also puts the PDBs in his or her installation directory.

**Comment by dfdffish** [comment]
![image](https://github.com/xStarbound/xStarbound/assets/93636612/130a8f25-adea-4840-8c86-6c1c6a0fdc0d)
I mean.. like this right? I'm using xserver so xclient didn't crash, only xserver logs that error. 

or you want me to do that in singleplayer?

**Comment by FezzedOne** [comment]
No need. The crash you reported is virtually identical to one reported by `@_kabut0_` on Discord last week. It affects both the MSVC and MinGW Windows builds and appears to be related to xStarbound not gracefully handling running out of memory on Windows when running entity scripts. I suspect this because when I was trying to reproduce this same error on the Linux build, I didn't get any crashes or errors aside from a one-time «out of memory» error that got logged for a Lua script.

**Comment by FezzedOne** [comment]
My current hypothesis is that the segfaults are being caused (somehow) by Windows-specific ABI compatibility shenanigans, especially around handling out-of-memory errors. @kblaschke has been working on «streamlining» the Windows build process with GitHub actions (as much as it can be); hopefully, this gets rid of any issues that might be caused by ABI incompatibility.

**Comment by dfdffish** [comment]
Looking forward to your good news! :D

**Comment by FezzedOne** [comment]
Thanks to @kblaschke, [automatic GitHub builds are out now](https://github.com/xStarbound/xStarbound/actions). I'm waiting until they get tested before releasing the next xStarbound update. I suggest testing these builds to see if the random crashes are still there. And don't forget to send any crash logs over to me.

**Comment by dfdffish** [comment]
So it took me a while to get xserver running. But I can't find a way to boot up xclient, it shows:
**"the procedure entry point SteamInternal_SteamAPI_Init could not be located in the dynamic link library"**
Do I need a specific version of steam_api64.dll for this?

Notifying me if I missed anything. Not familiar with C++ development. :/

**Comment by kblaschke** [comment]
The Steam SDK was updated, this DLL here should work, just download and replace:

https://github.com/xStarbound/xStarbound/raw/main/source/extern/steam/lib/windows/steam_api64.dll

I guess it's not being properly copied to the "dist" dir yet, same with the Discord SDK.

**Comment by dfdffish** [comment]
![QQ_1721280199139](https://github.com/user-attachments/assets/c5f190b3-a6ab-4037-a102-6bedb9587030)
And... this.

Should I just wait for a new build? This one doesn't really seem to work well on my PC.

**Comment by kblaschke** [comment]
This assertion looks more like an issue in the code itself, something going over the end of a string and the runtime library catching this issue. The code hasn't really changed much except for the RegEx fixes done by @FezzedOne lately, but those only affect Linux. I guess the new build has a few more checks as it uses a newer MSVC runtime, so this issue might just have been gone undetected as of now.

So you just ran xclient with the standard assets? I'll see if I can get my hands on a Windows PC with VS2022 installed and try to debug this one.

**Comment by dfdffish** [comment]
Yes. I copied xSBassets.pak directly from the previous version, and I also need to fill up some missing dependencies for this build to work:
**vcruntime140d.dll
vcruntime140_1d.dll
msvcp140d.dll
ucrtbased.dll**

Wait..You mean it's something related to MSVC runtime.. I'm starting to wonder if I also used wrong versions of these dll too... :(

The thing is, I should already have full sets of MSVC runtime installed. 
![QQ_1721289200087](https://github.com/user-attachments/assets/d0f528a5-56eb-4b69-ba92-4c1d0fe0acde)

But xclient told me that those runtime files were missing. So I went to download those files then put them into xsb-win64 folder.

**Comment by kblaschke** [comment]
You only have to install the latest x64 runtime redistributable here: https://aka.ms/vs/17/release/vc_redist.x64.exe
No need to copy any DLLs from the runtime into the SB dir. Not sure about the debug versions though, these might only be installed alongside the Windows SDK and not via the redist package. The release builds of xStarbound should only require the standard runtime libraries, not the (non-redistributable) debug ones.

What I meant was that the newer runtime the changed build uses (toolset v143) seemingly has more runtime diagnostics added, which kill the program if it does something illegal. In fact, xStarbound _does_ illegal things with C++ types, so that's why the runtime terminates the program - and for a good reason.

Still debugging the assertions, found two places where string iterators were used in a wrong way. I guess there may be a few more, but once those are fixed, the game should run and certainly be more stable as such issues can lead to memory corruption and thus causing crashes.

**Comment by dfdffish** [comment]
*sigh* Looks like quite a mess. Respect to you guys. I'll just wait patiently then. :)

**Comment by dfdffish** [comment]
So I got the newest test build and booted up the game successfully.

I have progressed through "apexmission1" for like 10 times(5 times in xserver, 5 times xclient-singleplayer) trying to reproduce the same crash, **but not even a single one!** So It seems that the random crashes are history now. I guess..? 

I'm not sure if those random crashes happened elsewhere are resolved too. But if this "apexmission1" one is fixed, so do the others. Right?

Also I haven't test in my actual save, still in a test environment.
